sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnReflector.java

```diff
@@ -91,1012 +36,268 @@
-            formatType(param), format(m), formatType(info.tokenFor(iActual, oActual))));
-      }
-
-      // Register the (now validated) context info
-      contextInfos[i - 1] = info;
-    }
-    return ImmutableList.copyOf(contextInfos);
-  }
-
-  /** Interface for invoking the {@code OldDoFn} processing methods. */
-  public interface DoFnInvoker<InputT, OutputT>  {
-    /** Invoke {@link OldDoFn#startBundle} on the bound {@code OldDoFn}. */
-    void invokeStartBundle(
-        DoFn<InputT, OutputT>.Context c,
-        ExtraContextFactory<InputT, OutputT> extra);
-    /** Invoke {@link OldDoFn#finishBundle} on the bound {@code OldDoFn}. */
-    void invokeFinishBundle(
-        DoFn<InputT, OutputT>.Context c,
-        ExtraContextFactory<InputT, OutputT> extra);
-
-    /** Invoke {@link OldDoFn#processElement} on the bound {@code OldDoFn}. */
-    public void invokeProcessElement(
-        DoFn<InputT, OutputT>.ProcessContext c,
-        ExtraContextFactory<InputT, OutputT> extra);
-  }
-
-  /**
-   * Implementation of {@link DoFnReflector} for the arbitrary {@link DoFn}.
-   */
-  private static class GenericDoFnReflector extends DoFnReflector {
-
-    private final Method startBundle;
-    private final Method processElement;
-    private final Method finishBundle;
-    private final List<AdditionalParameter> processElementArgs;
-    private final List<AdditionalParameter> startBundleArgs;
-    private final List<AdditionalParameter> finishBundleArgs;
-    private final Constructor<?> constructor;
-
-    private GenericDoFnReflector(
-        @SuppressWarnings("rawtypes") Class<? extends DoFn> fn) {
-      // Locate the annotated methods
-      this.processElement = findAnnotatedMethod(ProcessElement.class, fn, true);
-      this.startBundle = findAnnotatedMethod(StartBundle.class, fn, false);
-      this.finishBundle = findAnnotatedMethod(FinishBundle.class, fn, false);
-
-      // Verify that their method arguments satisfy our conditions.
-      this.processElementArgs = verifyProcessMethodArguments(processElement);
-      this.startBundleArgs = verifyBundleMethodArguments(startBundle);
-      this.finishBundleArgs = verifyBundleMethodArguments(finishBundle);
-
-      this.constructor = createInvokerConstructor(fn);
-    }
-
-    private static Collection<Method> declaredMethodsWithAnnotation(
-        Class<? extends Annotation> anno,
-        Class<?> startClass, Class<?> stopClass) {
-      Collection<Method> matches = new ArrayList<>();
-
-      Class<?> clazz = startClass;
-      LinkedHashSet<Class<?>> interfaces = new LinkedHashSet<>();
-
-      // First, find all declared methods on the startClass and parents (up to stopClass)
-      while (clazz != null && !clazz.equals(stopClass)) {
-        for (Method method : clazz.getDeclaredMethods()) {
-          if (method.isAnnotationPresent(anno)) {
-            matches.add(method);
-          }
+      if (param.isSubtypeOf(BoundedWindow.class)) {
```

Comment:
Replaced all of these subtyping checks with equality checks, for reasons discussed offline.